### PR TITLE
Ensure log group subscription filter resource has stable id

### PR DESCRIPTION
**Pull request by mweberxyz**
### What does this PR do?

This PR calls ensures the log filter subscription resource id does not change on subsequent cdk deploys.

### Motivation

Every time I deploy my app utilizing datadog-cdk-constructs, I see that the log subscription filters are deleted and re-added by CloudFormation. This due to references being passed in as the functionArn to generateForwaderConstructId, and the reference ID numbers are not stable across synth operations. See below for back-to-back cdk diff operations:

CleanShot 2021-04-15 at 20 59 10

### Testing Guidelines

An additional unit test was added that shows two identical -- but independent -- stacks produce the same name of the subscription filter, but a third stack with a differently named resource does not.

### Additional Notes

The implementation of the existing unit tests inlined the actual id generation algorithm, so I modified these tests to instead introspect the CDK data structures to confirm everything is working.

### Types of Changes

- [x] Bug fix
- [ ] New feature
- [ ] Breaking change
- [x] Misc (test refactoring)

### Check all that apply

- [x] This PR's description is comprehensive
- [ ] This PR contains breaking changes that are documented in the description
- [ ] This PR introduces new APIs or parameters that are documented and unlikely to change in the foreseeable future
- [ ] This PR impacts documentation, and it has been updated (or a ticket has been logged)
- [x] This PR's changes are covered by the automated tests
- [ ] This PR collects user input/sensitive content into Datadog
- [ ] This PR passes the integration tests (ask a Datadog member to run the tests)

### CODE REVIEW

**Comment by zARODz11z** [comment]
@mweberxyz Thanks for your contribution I will take a look

**Comment by mweberxyz** [comment]
Thanks @zARODz11z, to contrast with my screenshot in the description of the problem, here is the result of two consecutive `cdk diff` in my primary project when incorporating my changes. Note each run attempts to set the resource ID to the same value (and then future diff/deploy from that point forward would ensure the subscription filters are not re-created):

CleanShot 2021-04-16 at 15 37 12

**Comment by zARODz11z** [comment]
For context the reason why we have that subscription filter id logic is because we had issues with duplicate constructs in the same construct tree. So I may have to do some research/testing to see if these changes don't re-introduce that problem.

**Comment by zARODz11z** [comment]
I took a look and even manually tested your changes, they look solid. For the sake of getting a fresh look at it, I am going to wait until Monday to provide any feedback/approve these changes.

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:58]: I like how you refactored the tests but how can you be sure that index 0 of the `subscriptionFilters` array will always be the subscription filter associated with `nodeLambda` and subsequently how index 1 will be the filter associated with `pythonLambda`

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:80]: I noticed that the `Subscribes the same forwarder to two different lambda functions via separate addLambdaFunctions function calls` test uses `datadogCdk.addLambdaFunctions([nodeLambda]);` and `datadogCdk.addLambdaFunctions([pythonLambda]);` to add the subscriptions however this test directly uses `addForwarder`. Could you update the tests to use the `addForwarder` function instead of the `datadogCdk.addLambdaFunctions()` way? 

This way the convention in the file is the same and so the `addForwarder` set of tests really only test the `addForwarder` function and not all the other codepaths that come with using `datadogCdk.addLambdaFunctions()`

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:54]: see related [comment](https://github.com/DataDog/datadog-cdk-constructs/pull/30/files#r616051414)

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:55]: see related [comment](https://github.com/DataDog/datadog-cdk-constructs/pull/30/files#r616051414)

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:154]: see related [comment](https://github.com/DataDog/datadog-cdk-constructs/pull/30/files#r616051414)

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:155]: see related [comment](https://github.com/DataDog/datadog-cdk-constructs/pull/30/files#r616051414)

**Review by zARODz11z — CHANGES_REQUESTED** [review]
@mweberxyz Overall this is a great addition however I left some requests for changes and questions before merging.

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:28]: addLambdaFunctions -> addForwarder

File: test/index.spec.ts
Comment (zARODz11z) [inline test/index.spec.ts:88]: disregard this [comment](https://github.com/DataDog/datadog-cdk-constructs/pull/30/files#r616051414) for this test

File: test/index.spec.ts
Comment (mweberxyz) [reply test/index.spec.ts]: Good question. The answer was "it works today on my machine" but for sanity's sake I factored it out to look at the children of the lambda's themselves.

File: test/index.spec.ts
Comment (mweberxyz) [reply test/index.spec.ts]: I didn't want to get too deep into reworking the tests, but seems best to move all addForwarder specific tests to the forwarder.spec.ts, and factor out the "search the cdk nodes" logic to be reused by both index.spec and forwarder.spec

Let me know how you feel about the most recent commit!

**Comment by mweberxyz** [comment]
Hey there @zARODz11z let me know how you feel about the most recent commit. I think I understood what you were getting at.

I added an export to forwarder.ts to avoid redefining the constant in the tests, then  moved the tests that invoke addForwarder specifically to forwarder.spec and factored out the CDK tree introspection to use it both in index.spec and forwarder.spec.

File: test/test-utils.ts
Comment (zARODz11z) [inline test/test-utils.ts:14]: Is this because you are only ever expecting to get 1 subscription filter per call to `findDatadogSubscriptionFilters`

**Review by zARODz11z — COMMENTED** [review]
@mweberxyz  This looks good to me, I left 1 question.

File: test/test-utils.ts
Comment (mweberxyz) [reply test/test-utils.ts:14]: Correct, I am asserting that the baseConstruct only has a single subscription filter.

Because a future test might want to ensure that a given stack as the baseConstruct has 2 subscription filters (for example, with 2 forwarder arns) I think it makes sense to gate this with a typecheck -- any only assert that there is a single subscription filter if the baseConstruct is a Function. I will push this shortly.

Comment (mweberxyz) [reply test/test-utils.ts]: I pushed the change but actually.. hold off on reviewing. I will make this a bit clearer.

**Comment by mweberxyz** [comment]
Alrighty @zARODz11z I think it's as clear as possible at this point, let me know what you think.

**Review by zARODz11z — APPROVED** [review]
This looks good to me, thanks for the contribution!

**Pull request merged**
